src/hotspot/share/services/virtualMemoryTracker.cpp

```diff
@@ -523,29 +525,72 @@ bool VirtualMemoryTracker::remove_released_region(address addr, size_t size) {
     }
   }
 
-  VirtualMemorySummary::record_released_memory(size, reserved_rgn->flag());
-
-  assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
-  if (reserved_rgn->base() == addr ||
-      reserved_rgn->end() == addr + size) {
+  if (size <= reserved_rgn->size() && reserved_rgn->contain_region(addr, size)) {
+    // The requested chunk to be freed is smaller or same size as the real
+    // underlaying region and is fully contained within one (i.e. single region)
+    VirtualMemorySummary::record_released_memory(size, reserved_rgn->flag());
+    assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
+    if (reserved_rgn->base() == addr || reserved_rgn->end() == addr + size) {
       reserved_rgn->exclude_region(addr, size);
```

Comment:
This is unfortunately not named "shrink_region_from" or something similar :-(. I don't expect you to rename the function, to be clear :).

Reply:
I'm not 100% liking the name "shrink_region_from", but if we were to find one that we can all agree on, I can factor out that code into its own method.